service/history/replication/sequential_batch_queue.go

```diff
@@ -0,0 +1,151 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package replication
+
+import (
+	"sync"
+
+	"go.temporal.io/server/common/collection"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/metrics"
+	ctasks "go.temporal.io/server/common/tasks"
+)
+
+type (
+	SequentialBatchableTaskQueue struct {
+		id interface{}
+
+		sync.Mutex
+		taskQueue                    collection.Queue[TrackableExecutableTask]
+		lastTask                     TrackableExecutableTask
+		batchedIndividualTaskHandler func(task TrackableExecutableTask)
+		logger                       log.Logger
+		metricsHandler               metrics.Handler
+	}
+)
+
+func NewSequentialBatchableTaskQueue(
+	task TrackableExecutableTask,
+	batchedIndividualTaskHandler func(task TrackableExecutableTask),
+	logger log.Logger,
+	metricsHandler metrics.Handler) ctasks.SequentialTaskQueue[TrackableExecutableTask] {
+	return &SequentialBatchableTaskQueue{
+		id: task.QueueID(),
+
+		taskQueue: collection.NewPriorityQueue[TrackableExecutableTask](
+			SequentialTaskQueueCompareLess,
+		),
+		batchedIndividualTaskHandler: batchedIndividualTaskHandler,
+		logger:                       logger,
+		metricsHandler:               metricsHandler,
+	}
+}
+
+func (q *SequentialBatchableTaskQueue) ID() interface{} {
+	return q.id
+}
+
+func (q *SequentialBatchableTaskQueue) Peek() TrackableExecutableTask {
+	q.Lock()
+	defer q.Unlock()
+	return q.taskQueue.Peek()
+}
+
+// Add will try to batch input task with the last task in the queue. Since most likely incoming task
+// are ordered by task ID, we only try to batch incoming task with last task in the queue.
+func (q *SequentialBatchableTaskQueue) Add(task TrackableExecutableTask) {
+	q.Lock()
+	defer q.Unlock()
+	q.updateLastTask(task)
+
+	batchableTask, isBatchableTask := task.(BatchableTask)
+
+	// TODO: Once POC approved, will add unit test to test following cases thoroughly
+	// case: input task is not a batchable task or input task does not want to be batched: simply add task into the queue
+	if !isBatchableTask || !batchableTask.CanBatch() {
+		q.updateLastTask(task)
+		q.taskQueue.Add(task)
+		return
+	}
+
+	// case: If last task is nil, create a batchedTask and put into the queue
+	if q.lastTask == nil {
```

Comment:
line 81 will always make sure `q.lastTask != nil`?

Reply:
or you should simply remove line 81

Reply:
Right.. line 81 should be removed.. I changed the implementation. So it is no longer relavent.